### PR TITLE
Issues #9 (drainpipe-dev): Adds GitLab CI includes

### PR DESCRIPTION
**Pull request by justafish**
GitLab CI configurations for common tasks, as well as bootstrapping a Pantheon Multidev workflow. Also includes some additional scaffold files for setting up a Pantheon site.

## Testing Instructions

- [x] Create a new Pantheon sandbox site for Drupal 9, visit the development environment to install Drupal (any profile), then set the development mode to git
- [x] Create a new GitLab project (premium or self-hosted)
- [x] Install DDEV and Yarn
- [x] Create a new test project
```
composer create-project drupal/recommended-project drainpipe-test
cd drainpipe-test
```
- [x] Configure ddev (accepting the default options)
```
ddev config
ddev start
```
- [x] Add this fork
```
ddev composer config repositories.lullabot/drainpipe git https://github.com/lullabot/drainpipe.git
ddev composer config minimum-stability dev
```
- [x] Add the following to the extras field in `composer.json`
```
extra": {
    "drainpipe": {
        "gitlab": [
                "ComposerLockDiff",
                "Pantheon",
                "PantheonReviewApps"
            ]
        },
}
```
- [x] Require this project
```
ddev composer require lullabot/drainpipe:dev-justafish/drainpipe-dev-9/gitlab-ci-includes
```
- [x] Verify you have:
  - `.gitlab-ci.yml`
  - GitLab includes in `.drainpipe/gitlab`
  - `pantheon.yml`
  - `web/sites/default/settings.pantheon.php`
  - `.gitignore`
  - `.drainpipeignore`
- [x] Edit `web/sites/default/settings.php` and change the config export directory
```
$settings['config_sync_directory'] = '../config';
```
- [x] Install Drupal and export config
```
ddev drush site:install minimal --yes
ddev drush config:export --yes
```
- [x] Add the following to `web/sites/default/settings.php`
```
/**
 * Pantheon settings.
 */
if (!empty(getenv('PANTHEON_ENVIRONMENT'))) {
  include __DIR__ . '/settings.pantheon.php';
  if (!empty($_ENV['PANTHEON_ENVIRONMENT']) && !in_array($_ENV['PANTHEON_ENVIRONMENT'], ['live', 'test', 'dev'])) {
    $settings['container_yamls'][] = __DIR__ . '/services.pantheon.multidev.yml';
  }
}
```
- [x] In `.gitlab-ci.yml` change the environment variable values and set project CI/CD variables as appropriate (you'll need to generate a new SSH keypair and generate new terminus and gitlab access tokens). When adding variables to GitLab, make sure "Protect variable" is unchecked and "Mask variable" is checked (where possible)
 - [x] Run `git init`, add everything, commit it and push it to your GitLab repository
- [x] Start a new branch
- [x] Require a random package e.g. `ddev composer require whatthejeff/nyancat-phpunit-resultprinter`
- [x] Commit the changes, push and create a new merge reqest
- [x] Verify the MR description is updated with a table of composer changes
- [x] Verify the a Multidev environment is successfully created (will be linked below the merge request description)
- [x] Verify the browser performance job is successful (you won't get a report yet as it won't have run on the `main` branch)
- [x] Close the MR and verify the Stop job runs and removes the Multidev environment

### CODE REVIEW

**Comment by deviantintegral** [comment]
😮 so much! Looking forward to reviewing this. I've booked an hour tomorrow for this.

**Comment by justafish** [comment]
@deviantintegral :+1::+1::+1:
I've got more Pantheon integration in a new PR, so might be good to leave your sandbox site and GitLab project up once this is tested

**Comment by deviantintegral** [comment]
I get an error with this request. While I'm using a personal access token, it works if I run the curl command locally.

```
$ composer global exec -- "composer-lock-diff --md > composer_lock_diff.md"
Changed current directory to /root/.composer
$ curl --fail -H "PRIVATE-TOKEN: $GITLAB_ACCESS_TOKEN" https://gitlab.com/api/v4/projects/$CI_PROJECT_ID/merge_requests/$CI_MERGE_REQUEST_IID | jq '. | {id, iid, description}' > merge_request.json
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
curl: (22) The requested URL returned error: 404
Cleaning up project directory and file based variables
ERROR: Job failed: exit code 1
```

```
Using docker image sha256:657d9d3af244e77dda7b691b3e0574c03a46212c0e4483e5762f7b48ca89281c for webdevops/php-nginx:7.4-alpine with digest webdevops/php-nginx@sha256:d2a6e18e0a6b962c0e9ec1bc41eea1e43c52c060af0c1e4f651e039301c6eb03 ...
$ if [ "$CI_MERGE_REQUEST_EVENT_TYPE" == "merge_train" ]; then exit 0; fi
$ command -v ssh-agent >/dev/null || ( apk add --update openssh-client )
$ eval $(ssh-agent -s)
Agent pid 47
$ echo "$SSH_PRIVATE_KEY" | tr -d '\r' | ssh-add -
Error loading key "(stdin)": invalid format
Cleaning up project directory and file based variables
ERROR: Job failed: exit code 1
```

What's odd is if I copy the private key out and run the command locally, I'm able to add the key. Type is set to variable, scope is all, and protect variable on. Any ideas?

From a high level, I'm not sure defaulting to including `*` makes sense. We have lots of clients who aren't using Pantheon, so perhaps that makes sense to leave as a manual step to include the Pantheon jobs?

_I love_ these testing steps btw. They're 70% of the way towards the docs we'll want to write around this setup.

**Comment by deviantintegral** [comment]
> Install DDEV and Yarn

I think that's only needed inside ddev, which is automatic?

File: README.md [hunk 1]
```diff
@@ -37,50 +37,7 @@ Build tasks are provided as [Taskfiles](https://taskfile.dev/#/). A full list of
 available tasks can be shown by running `./vendor/bin/task --list` (or
 `ddev task --list` if you're using DDEV).
 
-### Drupal
```
Comment (deviantintegral) [inline README.md:40]: Why are we removing these docs?

**Review by deviantintegral — COMMENTED** [review]
I haven't done a line by line review which is probably a good idea at a later point.

**Comment by justafish** [comment]
@deviantintegral protect should be off/unchecked, otherwise it can't pass the secret to your branch

File: README.md (hunk 1 above)
Comment (justafish) [reply README.md:40]: they're not covering everything and get out of sync quickly, so better to just tell the user to run `--list` for now imo

File: scaffold/pantheon/pantheon.yml [hunk 2]
```diff
@@ -0,0 +1,7 @@
+api_version: 1
```
[the remark this answers is not in the document]
Comment (justafish) [reply scaffold/pantheon/pantheon.yml:1]: sadly all their linked examples are woefully out of date, but we could grab https://github.com/pantheon-systems/drupal-recommended/blob/default/pantheon.upstream.yml and rename it. We'll have to process it and turn `build_step: true` to `false`. I hesitate to do this though given they might suddenly change it and break something (ala `build_step`)

**Comment by rabbitlair** [comment]
@justafish, all fine for me, except for the item _Verify the MR description is updated with a table of composer changes_. I tried with a personal access token that has API scope, then another one with all available scopes, but it is not working on any case.

**Comment by rabbitlair** [comment]
Everything is working now. Approving!

**Review by rabbitlair — APPROVED** [review]

**Pull request merged**